Do not build tests unless needed, fix tests with introspection

https://bugs.gentoo.org/226345
https://bugzilla.gnome.org/show_bug.cgi?id=642624

---
--- a/tests/runtests.py
+++ b/tests/runtests.py
@@ -5,6 +5,8 @@ import glob
 
 import unittest
 
+# Some tests fail with translated messages.
+os.environ["LC_ALL"] = "C"
 
 # Load tests.
 if 'TEST_NAMES' in os.environ:
--- a/tests/Makefile.am
+++ b/tests/Makefile.am
@@ -1,7 +1,7 @@
 CLEANFILES =
-noinst_LTLIBRARIES = 
+check_LTLIBRARIES = 
 if ENABLE_INTROSPECTION
-noinst_LTLIBRARIES += libregress.la libgimarshallingtests.la
+check_LTLIBRARIES += libregress.la libgimarshallingtests.la
 
 nodist_libregress_la_SOURCES = $(GI_DATADIR)/tests/regress.c $(GI_DATADIR)/tests/regress.h
 libregress_la_CFLAGS = $(GIO_CFLAGS) $(PYCAIRO_CFLAGS)
@@ -46,7 +46,7 @@ gschemas.compiled: org.gnome.test.gschema.xml
 CLEANFILES += Regress-1.0.gir Regress-1.0.typelib GIMarshallingTests-1.0.gir GIMarshallingTests-1.0.typelib gschemas.compiled
 endif
 
-noinst_LTLIBRARIES += testhelper.la
+check_LTLIBRARIES += testhelper.la
 
 testhelper_la_CFLAGS = -I$(top_srcdir)/gobject -I$(top_srcdir)/glib $(PYTHON_INCLUDES) $(GLIB_CFLAGS)
 testhelper_la_LDFLAGS = -module -avoid-version
@@ -66,8 +66,6 @@ testhelper.la: $(testhelper_la_OBJECTS) $(testhelper_la_DEPENDENCIES)
 	test -L $@ || $(LN_S) .libs/$@ $@
 
 
-all: $(LTLIBRARIES:.la=.so)
-
 TEST_FILES_STATIC = \
 	test_gobject.py \
 	test_interface.py \
@@ -101,7 +99,7 @@ EXTRA_DIST = \
 EXTRA_DIST += $(TEST_FILES_STATIC) $(TEST_FILES_GI)
 
 clean-local:
-	rm -f $(LTLIBRARIES:.la=.so) file.txt~
+	rm -f $(check_LTLIBRARIES:.la=.so) file.txt~
 
 DBUS_LAUNCH=$(shell which dbus-launch)
 RUN_TESTS_ENV_VARS= \
@@ -111,7 +109,7 @@ RUN_TESTS_ENV_VARS= \
 RUN_TESTS_LAUNCH=$(RUN_TESTS_ENV_VARS) $(DBUS_LAUNCH) $(EXEC_NAME) $(PYTHON) $(srcdir)/runtests.py

 # run tests in separately to avoid loading static and introspection bindings in the same process
-check-local: $(LTLIBRARIES:.la=.so) Regress-1.0.typelib GIMarshallingTests-1.0.typelib gschemas.compiled
+check-local: $(check_LTLIBRARIES:.la=.so) @ENABLE_INTROSPECTION_TRUE@Regress-1.0.typelib GIMarshallingTests-1.0.typelib gschemas.compiled
 	TEST_FILES="$(TEST_FILES_STATIC)" $(RUN_TESTS_LAUNCH)
 	TEST_FILES="$(TEST_FILES_GI)" $(RUN_TESTS_LAUNCH)
 
